man/list.man
.\" 3.2
.\"  /*******************************************************************\
.\"   * Copyright (C) 1995 Lars J Aas <larsa@colargol.stud.idb.hist.no> *
.\"   * based on documentation by C J Needham <col@imdb.com> 1995,      *
.\"   * permission is granted by the authors to freely distribute       *
.\"   *                        providing no fee of any kind is charged. *
.\"  \*******************************************************************/
.\"
.TH LIST 1 "RELEASEDATE" " " "The Internet Movie Database vVERSIONNUMBER"
.SH NAME
list \- generates filmographies for people in the movie-database.
.SH SYNOPSIS
.B
list <searchmode> "name" [<searchmode> "name"] [<options>]
.SH DESCRIPTION
.B list
is a program that will search through the specified filmography lists
for information about some person, and output what it managed to find.
.SH OPTIONS
Options are divided into three groups:
.BR searchmode-options ,
.B output-options
and
.BR control-options .
.SS SEARCHMODES
.TP
.B \-name
This option will searches all of the filmography lists for information.
This is rather inefficient if you know which list the person will be on.
If you know which list, you should use one of the other options.
.TP
.B \-cast
Search both cast databases.
.TP
.B \-acr
Search actors database.
.TP
.B \-acs
Search actresses database.
.TP
.B \-dir
Search directors database.
.TP
.B \-write
Search writers database.
.TP
.B \-comp
Search composers database.
.TP
.B \-cine
Search cinematographers database.
.TP
.B \-edit
Search editors database.
.TP
.B \-prodes
Search production designers database.
.TP
.B \-costdes
Search costume designers database.
.TP
.B \-prdcr
Search producers database.
.TP
.B \-misc
Search miscellaneous filmographies database.
.SS EXAMPLES ON SEARCHMODES
list \-acr "Bogart, Humphrey"
.br
list \-acr "Biehn, Michael"
.br
list \-acs "Hepburn, Katharine"
.br
list \-dir "Cameron, James"
.SS OUTPUT-OPTIONS
.TP
.B \-yr
Add the year of release to each title, and sort chronologically.
.TP
.B \-yru
Add the year of release, but leave unsorted.
.TP
.B \-mrr
Adds movie ratings report information to each title.
.TP
.B \-smrr
Adds adds movie ratings report information, and sort by average rating.
.TP
.B \-vmrr
Adds movie ratings report information, and sort by number of votes.
.TP
.B \-chr
For cast lists, adds character names in []'s where available.
.TP
.B \-aka
Add alternative titles to the output.
.TP
.B \-bio
Include information from biographies list and formats output neatly.
.TP
.B \-full
Combines
.BR \-yr ,
.BR \-chr ,
.B \-aka
and
.BR \-bio .
.SS EXAMPLES ON OUTPUT\-OPTIONS
list \-acr "Grant, Cary" \-yr \-chr
.br
list \-dir "Hitchcock, Alfred" \-smrr
.SS CONTROL\-OPTIONS
.TP
.B \-m
Movies only: ignores TV\-series and mini\-series.
.TP
.B \-i
Case sensitive search: normally searches are not case sensitive.
.TP
.B \-genre <s>
Only include titles with a genre equal to
.BR <s> .
.TP
.B \-time <s>
Only include titles with a time equal to
.BR <s> .
.TP
.B \-prodco <s>
Only include titles with a production company equal to
.BR <s> .
.TP
.B \-cert <s>
Only include titles with a certificate equal to
.BR <s> .
.TP
.B \-mix <s>
Only include titles with a sound mix equal to
.BR <s> .
.TP
.B \-cntry <s>
Only include titles with a country equal to
.BR <s> .
.TP
.B \-rel <s>
Only include titles with a release date equal to
.BR <s> .
.TP
.B \-loc <s>
Only include titles with a location equal to
.BR <s> .
.TP
.B \-tech <s>
Only include titles with technical info equal to
.BR <s> .
.SS EXAMPLES ON CONTROL\-OPTIONS
list -dir "Hitchcock, Alfred" -prodco Selznick
.br
list -dir "Hitchcock, Alfred" -mix Silent
.br
list -dir "Lang, Fritz" -genre Film-Noir -full -color "Black and White"
.SH Substring and Approximate Matching
Ordinarily, the program requires an exact match in order to locate the 
information on a person. In some cases the exact spelling may not be known.
There are two classes of options which can help locate the specific person:
the first is substring matching, which treats the name you provided with the 
database option as a substring. The option to trigger this is
.BR \-s .
.PP
The second class is approximate matching, which enables names with a limited 
number of character differences to still match. The options to trigger this 
are \-1 \-2 \-3 \-4 \-5 \-6 \-7 \-8 and \-9 where the larger the number, the
larger the number of differences allowed for a match. 
.PP
The use of any of these options automatically switches on the raw output
mode (\-r) since you will need to see which names have been matched.
.SS EXAMPLES ON Substring and Approximate Matching"
list \-cast "Barrymore" \-s \-full
.br
list \-acr "rGant, Cary" \-2 \-full
.SH Multiple Name Queries
Any number of database and name options may be combined in the same query
to generate a list of movies that all the people worked on together. 
.SS EXAMPLES ON "Multiple Name Queries"
list \-dir "Hitchcock, Alfred" \-acr "Grant, Cary" 
.br
list \-acs "Hepburn, Katharine" \-acr "Tracy, Spencer" \-yr
.br
list \-acr "Sheen, Charlie" \-acr "Estevez, Emilio"


docs/ChangeLog
-------Version 3.4 Released---------------------

Sun Mar 29 22:40:37 BST 1998

 * biographies list offset fields increased to 4 bytes following overflow
 * various constants updated

-------Version 3.3c Released---------------------

Sat Apr 26 06:51:19 BST 1997

 * added support for crew completion list
 * added support for distributors list
 * adapted mkdbs for more robust processing of filmography lists
 * added support for new business list tag

-------Version 3.3b Released---------------------

Sat Mar 15 08:09:50 GMT 1997

 * added support for ISO-Latin-1

-------Version 3.3 Released---------------------

Wed Oct 16 22:28:54 BST 1996

 * added support for AKA attributes
 * added support for writers ordering
 * moved misc section to after case in title output
 * added support for business list
 * added support for laserdisc list
 * fixed tagline bug
 * diplays URL of web page for name/title

-------Version 3.2h Released---------------------

Tue Oct  8 14:24:21 BST 1996

 * added support for special effects companies list
 * updated various system constants
 * added new movie link types 'edited into' & 'edited from'
 * added extra checks to avoid title/attribute key file corruption in mkdbs.c

-------Version 3.2f Released---------------------

Mon Jun 24 22:04:46 GMT 1996

 * added support for language list

-------Version 3.2e Released---------------------

Sat Mar 16 08:09:39 GMT 1996

 * added support for alternate versions list
 * added bio keys IT & AT

-------Version 3.2d Released---------------------

Tue Aug 15 23:23:47 GMT 1995

 * switched TitleID to unsigned long to handle >65K titles
 * fixed bug in display of literature and genre data
 * added bio key WN

-------Version 3.2c Released---------------------

Sun May 14 23:00:14 GMT 1995

 * raised MAXFILMOGRAPHIES constant to handle larger no. of entries

-------Version 3.2 Released---------------------

Sun Jan 29 22:40:08 GMT 1995

 * added full support for title searching without year/TV/V info

Tue Jan  3 23:00:45 GMT 1995

 * replaced 'biodefs' file
 * fixed bug when looking for common titles with people using -name

Mon Jan  2 21:52:08 GMT 1995

 * log file support
 * better separation of search/display code

Wed Dec 14 21:43:57 GMT 1994

 * fixed bug in 'title' for miscellaneous section
 * added support for the movie links list

Tue Dec 13 22:06:44 GMT 1994
 
 * noted when cast in credits output
 * fixed name ordering problem for names with (I)/(II) marker
 * display Actor/Actress in list -name where appropriate

Wed Dec  7 21:19:19 GMT 1994

 * added support for tag lines & complete cast list
 * put spaces around "." in cast results
 * standardized number of dashes printed out in 'title' / 'list'

Mon Dec  5 21:19:42 GMT 1994

 * multi-country productions displayed on same line

Wed Nov 30 18:14:16 GMT 1994

 * switched to new IMDB license
 * updated 'lfetch' to get new lists
 * fixed "No plot summaries message"
 * simplified MRR processing in mkdb
 * added support for locations & release date lists
 * switched 'Reviews' marker for clarity
 * added support for technical & literature lists

Mon Oct 31 21:42:03 GMT 1994
 
-------Version 3.1 Released---------------------

Mon Oct 17 19:04:26 GMT 1994
 
 * 3.1 code frozen
 * minor bug fixes with lindex

Sun Oct  9 18:37:14 GMT 1994

 * added support for countries list

Tue Oct  4 21:27:53 GMT 1994

 * added support for cast ordering info

Sun Oct  2 23:02:59 GMT 1994

 * added support for color info and sound mix lists
 * added constraint options to 'list'
 * standardized list options

Wed Sep 28 22:22:56 GMT 1994

 * re-ordered title info printing plus separated genres list
 * print genres on one line

Mon Sep 26 21:53:47 GMT 1994

 * added support for production companies list

Sun Sep 25 22:02:07 GMT 1994

 * added constraint options to 'lindex'

Tue Sep  6 20:09:52 GMT 1994

 * added support for certificates & genres list
 * fixed mkballot bug with -ar option

Tue Aug 23 20:08:04 GMT 1994

 * added support for miscellaneous filmographies list

Tue May 24 19:13:47 GMT 1994

 * moviedb-3.0 completed

Thu May 19 22:35:12 GMT 1994

 * switched to merged character names scheme

Tue May 17 21:09:45 GMT 1994

 * new 'mkdb' eliminating need for large array to hold name keys
 * removed duplicate aka titles in 'title -s -aka' queries

Wed May 11 22:11:59 GMT 1994

 * made upperCase a macro for increased speed
 * switched to Phil's new Makefiles

Wed May  4 17:46:08 GMT 1994

 * mapped case of titles/names to match key files
 * added support for running times list
 * lguide error messages

Tue May  3 21:38:30 GMT 1994
 
 * added support for Phil's imoviedb package
 * expanded compression support

Thu Apr 28 19:51:52 GMT 1994

 * finished rewrite to use new database representation

-------Version 2.9 Released---------------------

Wed Mar 16 21:39:40 GMT 1994

 * added support for producers list
 * added support for soundtracks list
 * added support for generic aka lists

Tue Mar 15 23:47:46 GMT 1994

 * replaced -crazy and -triv options with -trivia (also includes goofs & quotes)
 * added support for quotes list
 * added support for goofs list
 * handled -full error messages correctly
 * correct conditions for printing of missing plot summaries message
 * fixed lindex to search all lists when no list options given
 * -mrr no longer implies -m

Mon Mar 14 21:48:21 GMT 1994

 * new version of check-for-program
 * passed $MAKE variable to make subprocesses
 * lfetch modified to match full filenames
 * added 'keepchars' option for CLEANLISTS

-------Version 2.8 Released---------------------

Wed Dec  1 21:58:21 GMT 1993

 * man pages added to package
 * LaTeX documentation added to package

Sat Nov 27 20:53:40 GMT 1993

 * 2.8 code finished!

Thu Nov 25 18:17:00 GMT 1993

 * ported lseen to C

Tue Nov 23 19:16:48 GMT 1993

 * added gendb -votes option

Mon Nov 22 23:54:58 GMT 1993

 * ported lformat to C
 * removed yrf.c file

Wed Nov 17 22:36:44 GMT 1993


src/list.c
/*============================================================================
 *
 *  Program: list.c 
 *
 *  Version: 3.7
 *
 *  Purpose: search databases for filmographies
 *
 *  Author:  C J Needham <col@imdb.com>     
 *
 *  Copyright (c) 1996-1999 The Internet Movie Database Inc.
 *
 *  Permission is granted by the copyright holder to distribute this program
 *  is source form only, providing this notice remains intact, and no fee
 *  of any kind is charged. This software is provided as is and there are no 
 *  warranties, expressed or implied.
 * 
 *  Options:
 *    -cast <name>  search both cast databases
 *     -acr <name>  search actors databases
 *     -acs <name>  search actress databases
 *     -dir <name>  search directors database
 *   -write <name>  search writers database
 *    -comp <name>  search composers database
 *    -cine <name>  search cinematographers database
 *    -edit <name>  search editors database
 *  -prodes <name>  search production designers database
 * -costdes <name>  search costume designers database
 *   -prdcr <name>  search producers database
 *    -misc <name>  search miscellaneous database
 *    -name <name>  search all databases
 *
 *    -mrr        add movie ratings report info to results
 *    -smrr       add movie ratings report info & sort by ratings
 *    -vmrr       add movie ratings report info & sort by votes
 *
 *    -yru        add year of release info
 *    -yr         add year of release info & sort chronologically
 *
 *    -chr        add character name info (if available)
 *    -bio        add data from biographies database
 *    -aka        add aka data for each title
 *    -full       shorthand for -bio -yr -chr -aka
 *
 *    -m          movies only, ignore TV-series
 *    -s          treat the name search as a substring and return all matches
 *    -i          case sensistive search for substring
 *
 *    -1..-9      use approximate matching, max difference in range 1 to 9
 *
 *   -genre <str>  limit search by genre
 *   -keyword <str> limit search by keyword
 *    -cert <str>  limit search by certificate
 *    -time <str>  limit search by running time
 *  -prodco <str>  limit search by production company
 *   -color <str>  limit search by color information
 *     -mix <str>  limit search by sound mix
 *   -cntry <str>  limit search by country of production
 *     -rel <str>  limit search by release date
 *     -loc <str>  limit search by location
 *    -tech <str>  limit search by technical info
 *    -lang <str>  limit search by language info
 *
 *============================================================================
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "log.h"
#include "moviedb.h"
#include "dbutils.h"
#include "filmography.h"
#include "display.h"

#define LIST_USAGE1 "usage: list [-cast|-acr|-acs|-dir|-write|-comp|-cine|-edit|-prodes|-costdes|"
#define LIST_USAGE2 "            -prdcr|-misc|-name <name>]+ [-full] [-mrr|-smrr|-vmrr] [-yr|-yru]"
#define LIST_USAGE3 "            [-chr] [-bio] [-aka] [-m -s -i] [-s|-1|-2|-3|-4|-5|-6|-7|-8|-9]"
#define LIST_USAGE4 "            [-genre <s>] [-keyword <s>] [-time <s>] [-prodco <s>] [-cert <s>]"
#define LIST_USAGE5 "            [-mix <str>] [-cntry <s>] [-loc <s>] [-rel <s>]"
#define LIST_USAGE6 "            [-tech <s>] [-lang <s>]"

int main ( int argc, char **argv )
{
  struct nameSearchRec *combined = NULL, *nchain = NULL ;
  struct nameSearchOptRec options = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 } ;
  struct searchConstraints constraints ;
  int    err = FALSE ;
  int    i, j, okopt ;
  int    tidy = FALSE ;
  int    fullopt = FALSE ;

  logProgram ( argv [ 0 ] ) ;
  initialiseConstraints ( &constraints ) ;

  if ( argc == 1 )
    err = TRUE ;

  for ( i = 1 ; i < argc ; i++ )
    if ( strcmp ( "-i", argv [ i ] ) == 0 )
      options . casesen = TRUE ;
    else if ( strcmp ( "-m", argv [ i ] ) == 0 )
      options . mvsonly = TRUE ;
    else if ( strcmp ( "-aka", argv [ i ] ) == 0 )
      options . akaopt = TRUE ;
    else if ( strcmp ( "-mrr", argv [ i ] ) == 0 )
      options . mrropt = MRR ;
    else if ( strcmp ( "-smrr", argv [ i ] ) == 0 )
      options . mrropt = SMRR ;
    else if ( strcmp ( "-vmrr", argv [ i ] ) == 0 )
      options . mrropt = VMRR ;
    else if ( strcmp ( "-yr", argv [ i ] ) == 0 )
      options . yropt = YR ;
    else if ( strcmp ( "-yru", argv [ i ] ) == 0 )
      options . yropt = YRU ;
    else if ( strcmp ( "-chr", argv [ i ] ) == 0 )
      options . chopt = TRUE ;
    else if ( strcmp ( "-s", argv [ i ] ) == 0 )
    {
      options . substring = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-bio", argv [ i ] ) == 0 )
    {
      options . biopt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-full", argv [ i ] ) == 0 )
    {
      options . yropt = YR ;
      options . biopt = TRUE ;
      options . chopt = TRUE ;
      options . akaopt = TRUE ;
      fullopt = TRUE ;
      tidy = TRUE ;
    }
    else if ( ( strlen ( argv [ i ] ) == 2 ) && ( argv [ i ] [ 1 ] >= '1' ) && ( argv [ i ] [ 1 ] <= '9' ) )
      options . approx = argv [ i ] [ 1 ] - '0' ;
    else
    {
      okopt = FALSE ;
      for ( j = 0 ; j < NO_OF_SEARCH_OPTS ; j++ )
	if ( strcmp ( filmographyOptions [ j ] . option, argv [ i ] ) == 0 || strcmp ( filmographyOptions [ j ] . oldOption, argv [ i ] ) == 0 )
        {
	  okopt = TRUE ;
          if ( ++i < argc )
          {
            nchain = addNameSearchRec ( nchain, argv [ i ], j ) ;
            if ( j == GLOBAL_NAME_SEARCH )
              tidy = TRUE ;
          }
          else
          {
            err = TRUE ;
	    break ;
          }
	}
      if ( ! okopt )
      {
        for ( j = 0 ; j < NO_OF_TITLE_INFO_LISTS ; j++ )
          if ( strcmp ( titleInfoDefs [ j ] . option, argv [ i ] ) == 0 )
          {
            okopt = TRUE ;
            if ( ++i < argc )
              (void) strcpy ( constraints . titleInfoString [ j ], argv[i] ) ;
            else
              err = TRUE ;
          }
      }
      if ( ! okopt )
      {
         (void) fprintf ( stderr, "list: unrecognised option %s\n", argv[i] ) ;
         err = TRUE ;
      }
    }

  if ( err || nchain == NULL )
    moviedbUsage ( LIST_USAGE1, LIST_USAGE2, LIST_USAGE3, LIST_USAGE4, LIST_USAGE5, LIST_USAGE6 ) ;

  if ( nchain -> next != NULL )
  {
    if ( options . substring )
      moviedbError ( "list: -s not available with more than one name search" ) ;
    else if ( fullopt )
      moviedbError ( "list: -full not available with more than one name search" ) ;
    else if ( options . biopt )
      moviedbError ( "list: -bio not available with more than one name search" ) ;
    else if ( options . chopt )
      moviedbError ( "list: -chr not available with more than one name search" ) ;
  }

  if ( options . approx && options . substring )
    moviedbError ( "list: -s not available with approximate matching" ) ;

  if ( nchain != NULL && nchain -> next != NULL )
  {
    addNameChainOpts ( nchain, options ) ;
    processFilmographySearch ( nchain ) ;
    constrainFilmographySearch ( &constraints, nchain ) ;
    combined = combineListResults ( nchain ) ;
    if ( combined != NULL )
    {
      sortListResults ( combined ) ;
      displayNameSearchResults ( combined, FALSE ) ;
      freeNameSearchChain ( combined ) ; 
    }
  }
  else
  { 
    addNameChainOpts ( nchain, options ) ;
    processFilmographySearch ( nchain ) ;
    constrainFilmographySearch ( &constraints, nchain ) ;
    sortListResults ( nchain ) ;
    displayNameSearchResults ( nchain, tidy ) ;
  } 

  freeNameSearchChain ( nchain ) ; 

  return ( 0 ) ;
}
